### DISCUSSION TITLE
scrollIntoView options as parameters in the Scroll Interaction

### DISCUSSION THREAD

**Discussion by jotacoroe2y** [serenity-js/serenity-js#2041] [Show and tell]
## Description
Hi again,

in one of our projects, we usually need to scroll to an element but placing it into the middle of the viewport.
This is needed specially when working with viewports of portable devices, where there's usually both sticky header and footer.
So we usually faced situations where the click was intercepted by another element (these sticky elements in most of the cases), so our tests failed.

The current implementation of the [Scroll interaction](https://serenity-js.org/api/web/class/Scroll/) doesn't support the [scrollIntoView options](https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView), so we implemented the following to have the chance and choose the most suitable scroll depending on the situation:
```typescript
import { Answerable, Interaction } from '@serenity-js/core';
import { PageElement, Scroll as SerenityScroll } from '@serenity-js/web';
import { TIME } from 'constants/datetime';

const defaultScrollOptions: ScrollIntoViewOptions = {
  block: 'center',
  inline: 'center',
};

export const Scroll = {
  intoView: (scrollOptions = defaultScrollOptions) => ({
    to: (element: Answerable<PageElement<WebdriverIO.Element>>) =>
      Interaction.where(
        `#actor does a customized scroll to the element ${element.toString()}`,
        async (actor) => {
          const nativeElement = await actor
            .answer(element)
            .then((element) => element.nativeElement());

          await nativeElement.waitForExist({ timeout: TIME.fiveSecsInMs });
          await nativeElement.scrollIntoView(scrollOptions);
          await nativeElement.waitForDisplayed({ timeout: TIME.fiveSecsInMs });
          await browser.pause(TIME.oneSecInMs);
        },
      ),
  }),
  to: (pageElement: Answerable<PageElement>) => SerenityScroll.to(pageElement),
};
```

When calling the Interaction, it's pretty similar as calling the default Scroll interaction:
```typescript
Scroll.intoView({
      block: 'center',
      inline: 'center',
      behavior: 'smooth',
    }).to(myElement),
```
---

 ## Pros ✅ 
This is currently a valid solution for us because our setup consists basically on [this](https://github.com/serenity-js/serenity-js-cucumber-webdriverio-template) (Serenity-js, Cucumber, WebdriverIO).
Actually, all the methods used comes directly from WebdriverIO ([waitForExist](https://webdriver.io/docs/api/element/waitForExist/), [scrollIntoView](https://webdriver.io/docs/api/element/scrollIntoView/), [waitForDisplayed](https://webdriver.io/docs/api/element/waitForDisplayed/)). As it can be seen, the `scrollIntoView` method from WebdriverIO support these `scrollIntoViewOptions`.

## Cons ❌ 
- In the moment that we consider to stop using WebdriverIO, could be more complicated or even impossible to maintain this feature.
- If we want to use this interaction with an element, we cannot use the usually way to define a PageElement:
```typescript
const myElement = PageElement.located(
    By.css(`.my-element`),
  ).describedAs(`my element`),
```
But we need to type it as a WebdriverIO element:
```typescript
const myElement = PageElement.located<WebdriverIO.Element>(
    By.css(`.my-element`),
  ).describedAs(`my element`),
```

## Requesting 🙏 
> Could be possible to have scrollIntoView options implemented in the default Scroll interaction?

**Comment by Himanshu21git** [comment]
@jotacoroe2y I faced a similar problem. You may refer [this issue](https://github.com/serenity-js/serenity-js/issues/2000).

**Reply by jotacoroe2y** [reply]
absolutely. Thanks a lot @Himanshu21git , even doing a previous search before opening this issue, I didn't find yours.
